src/backend/drm/output.rs

```diff
@@ -0,0 +1,908 @@
+                    if let Err(err) =
+                        compositor.set_format(allocator.clone(), current_format, [DrmModifier::Invalid])
+                    {
+                        tracing::warn!(?err, "failed to set new format");
+                        continue;
+                    }
+
+                    render_elements.submit_composited_frame(&mut *compositor, renderer)?;
+                }
+
+                let compositor = write_guard.get_mut(crtc).unwrap().get_mut().unwrap();
+                match compositor.use_mode(mode) {
+                    Ok(_) => render_elements.submit_composited_frame(&mut *compositor, renderer)?,
+                    Err(err) => {
+                        // try to reset format
+
+                        for compositor in write_guard.values_mut() {
+                            let compositor = compositor.get_mut().unwrap();
+
+                            let current_format = compositor.format();
+                            if let Err(err) = compositor.set_format(
+                                allocator.clone(),
+                                current_format,
+                                renderer_formats
+                                    .iter()
+                                    .filter(|f| f.code == current_format)
+                                    .map(|f| f.modifier),
+                            ) {
+                                tracing::warn!(?err, "failed to reset format");
+                                continue;
+                            }
+
+                            render_elements.submit_composited_frame(&mut *compositor, renderer)?;
+                        }
+
+                        return Err(DrmOutputManagerError::Frame(err));
+                    }
+                }
+            }
+            Err(err) => return Err(DrmOutputManagerError::Frame(err)),
+        };
+    }
+
+    Ok(())
+}
+
+/// Set of render elements for a set of outputs managed by an [`DrmOutputManager`].
+///
+/// A few methods of the [`DrmOutputManager`] and [`DrmOutput`] might need to do
+/// commits to multiple surfaces to satisfy bandwidth constraints. To not render
+/// a black screen this struct can be populated with screen contents to be used
+/// when such an operation is required. Outputs not provided via
+/// [`DrmOutputRenderElements::add_output`] will fallback to black.
+#[derive(Debug)]
+pub struct DrmOutputRenderElements<R, E>
+where
+    E: RenderElement<R>,
+    R: Renderer + Bind<Dmabuf>,
+    <R as Renderer>::TextureId: Texture + 'static,
+    <R as Renderer>::Error: Send + Sync + 'static,
+{
+    render_elements: HashMap<crtc::Handle, (Vec<E>, Color32F)>,
+    _renderer: PhantomData<R>,
+}
+
+impl<R, E> Default for DrmOutputRenderElements<R, E>
+where
+    E: RenderElement<R>,
+    R: Renderer + Bind<Dmabuf>,
+    <R as Renderer>::TextureId: Texture + 'static,
+    <R as Renderer>::Error: Send + Sync + 'static,
+{
+    fn default() -> Self {
```

Comment:
While not super important we should consider adding a `with_capacity` function allowing to pre-allocate the intern hash map.